eluthia/zipapp_script.py

```diff
@@ -0,0 +1,157 @@
+import os
+import time
+
+APP_PATH = os.getcwd() # We actually have to use this instead of abspath to get the correct value
+
+HISTORY_PATH = '/var/lib/badtrack/history' #TODO: use environment variable.
+UNPACK_PATH = f'{APP_PATH}/zip_unpack' #TODO: This could also be an environment variable?
+NEW_DEB_PATH = f'{UNPACK_PATH}/new_debs'
+OLD_DEB_PATH = f'{UNPACK_PATH}/old_debs'
+
+
+def get_version(package_name):
+    """
+    Verifies that a package is installed, and get it's version.
+    Input:
+        string package_name - the name of the package that should be checked
+    Output:
+        string version - version of the package
+            NoneType None denotes that the package is not installed.
+    """
+    try:
+        # Get package version number by parsing dpkg -s.
+        return str(subprocess.check_output(['dpkg', '-s', package_name]))\
+                    .split('Version: ')[1]\
+                     .split('\\n')[0]
+
+    except subprocess.CalledProcessError: # This happens if the package is not installed. I could not find a way to check if packages are installed without causing an exception.
+        return None
+
+
+def unpack_zip(zip_path):
+    """
+    Unpacks the packages in the archive into NEW_DEB_PATH. Also extracts information about the packages.
+    Input:
+        string zip_path - path to the archive to be extracted (this should always be os.path.dirname(__file__), but it's a parameter for modularity)
+    Output:
+        dict new_packages - new packages to be installed and their versions. Format {package_name: version}
+        dict old_packages - old packages to be replaced and their versions. Format: {package_name: version}
+    """
+    def parse_package_names(namelist):
+        #TODO: Eventually, packages will be named with their version included. Once this change is made, this function will be properly implemented.
+        packages = {}
+        for item in namelist:
+            if not item.endswith('.deb'): # Guard clause, makes us skip __main__.py or any other non-package files that might be added in future
+                continue
+
+            #TODO: Here we need to split the name into version and package name.
+            version = None
+            name = item[:-4] # removes .deb from package name
+
+            packages[name] = version
+        return packages
+
+    with zipfile.ZipFile(zip_path) as z:
+
+        new_packages = parse_package_names(z.namelist())
+        old_packages = {package: get_version(package) for package in new_packages.keys() if get_version(package) is not None}
+
+        for package in new_packages.keys():
+            z.extract(f'{package}.deb', path=NEW_DEB_PATH) # Extract the packages to NEW_DEB_PATH
+
+    return new_packages, old_packages #I'm not 100% sure if I want to return these from unpack_zip(), but I like that it means we only need to open the zip once.
+
+
+def repack(package_name, version):
+    subprocess.run(['dpkg-repack', package_name])
```

Comment:
Use `cwd` argument on `run`, so that you don't have to use `os.getcwd` for `APP_PATH`, giving freedom to `APP_PATH` to be anywhere.